text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,403 @@
+
+We found that [some crates use `Deref` to express an is-a not a has-a relationship](ttps://gist.github.com/davidhewitt/d0ed031fb05f6db98ee249ae089b268e) and so, ideally, might have preferred the option of setting up `Deref` and `self` candidacy separately. But, on discussion, we concluded that traits would be a better way to model those relationships.
+
+## Explore both `Receiver` and `Deref` chains while identifying method candidates
+
+We could modify the method search algorithm to explore both `Deref` and `Receiver` targets when identifying method candidates. This would avoid breaking compatibility, yet would give the desired flexibility for folks who wish to implement `Receiver` but not `Deref`.
+
+We don't think this is such a good option because:
+
+* It's more confusing for users;
+* It could lead to a worst-case O(n^2) number of method candidates to explore (though possibly this could be limited to O(2n) if we added restrictions);
+* It's a more invasive change to the compiler;
+* We don't know of any use-cases which the `Receiver<Target=T>` and blanket implementation for `Deref` do not allow.
+
+If some use-case presents itself where a type _must_ implement `Deref` but not `Receiver`; or a use-case presents itself where `Deref` and `Receiver` _must_ have different `Target`s then we will have to consider this more complex option.
+
+## Generic parameter
+
+Change the trait definition to have a generic parameter instead of an associated type. There might be permutations here which could allow a single smart pointer type to dispatch method calls to multiple possible receivers - but this would add complexity, no known use case exists, and it might cause worst-case O(n^2) performance on method lookup.
+
+## Do not enable for pointers
+
+It would be possible to respect the `Receiver` trait without allowing dispatch onto raw pointers - they are essentially independent changes to the candidate deduction algorithm.
+
+We don't want to encourage the use of raw pointers, and would prefer rather that raw pointers are wrapped in a custom smart pointer that encodes and documents the invariants. So, there's an argument not to add the raw pointer support.
+
+However, the current unstable `arbitrary_self_types` feature provides support for raw pointer receivers, and with years of experience no major concerns have been spotted. We would prefer not to deviate from the existing proposal more than necessary. Moreover, we are led to believe that raw pointer receivers are quite important for the future of safe Rust, because stacked borrows makes it illegal to materialize references in many positions, and there are a lot of operations (like going from a raw pointer to a raw pointer to a field) where users don't need to or want to do that. We think the utility of including raw pointer receivers outweighs the risks of tempting people to over-use raw pointers.
```

Comment:
Doesn't that mean that adding methods on `*const T`, like [`add()`](https://doc.rust-lang.org/std/primitive.pointer.html#method.addr), is a breaking change? How has this been dealt with so far?

Reply:
Good point. There are various Rust standard library types - `Box`, `Arc` etc. - where by policy no new methods are added because it would be a breaking change. Instead, associated functions are added.

If arbitrary self types is enabled (either the existing nightly unstable version, or our slightly tweaked version here) then that policy would need to be extended to the raw pointer types as well.

The Rust community therefore needs to decide which is higher priority:
* being able to add new methods to raw pointer types; or
* being able to receive method calls by raw pointer.

Unless someone can see some workaround or compromise?

@Manishearth hello, I think you were especially keen that arbitrary self types continues to support raw pointers. Do you have any views?

(Personally I think this might be a good argument to enable arbitrary self types _without_ raw pointers, and therefore to require people to write their own newtype wrappers if they want method dispatch).

Reply:
(there's a little bit more discussion of this in [this comment thread below](https://github.com/rust-lang/rfcs/pull/3519/files#r1379378763) which I'm going to close to avoid duplication)

Reply:
I wanted to reply to [your comment](https://github.com/rust-lang/rfcs/pull/3519#discussion_r1380057062) about changing priority: I think this new problem is a much better problem. The original issue is "the pointer type `*const T` can't add any new methods ever", the new issue is "crate A can add `foo(*const Self)` methods, except that if they're from the set of known `*const T` methods, this is a major semver change". This is a bit surprising but highly manageable.

What worries me more is the high confusion cost of making method resolution more complicated.

Reply:
I'm still thinking this through but I wanted to note that we Rust doesn't current shadow things invisibly; instead it gives an error message if there's ambiguity.

* For arbitrary self types which are not raw pointers: we [show the ambiguity as an error](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=218cc52ca6d8c0b584c3fa06f77b234f#:~:text=Permalink%20to%20the%20playground).
* For arbitrary self types which are raw pointers, we [show similar errors if there are ambiguities, and we even warn if there are known upcoming future ambiguities](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ea541e54f00fa15729af54f7d40668f5):
```
   = warning: once this associated item is added to the standard library, the ambiguity may cause an error or change in behavior!
   = note: for more information, see issue #48919 <https://github.com/rust-lang/rust/issues/48919>
   = help: call with fully qualified syntax `MyType::addr(...)` to keep using the current method
   = help: add `#![feature(strict_provenance)]` to the crate attributes to enable `ptr::const_ptr::<impl *const T>::addr`
   = note: `#[warn(unstable_name_collisions)]` on by default
```

This RFC should be more explicit somewhere about the existence of these errors and warnings, especially in the Method Shadowing section - I'll adjust.

Of course, this doesn't solve the issue here: if Rust adds a new `<raw pointer>::repaint()` method, that will cause downstream crates to cease building if they had:

```rust
struct Bedroom;
impl Bedroom {
  fn repaint(self: *const Bedroom) {}
}
```

@Nadrieril 's proposal, as I understand it, is that if method resolution does result in ambiguities in such a situation, where raw pointers are involved, we don't show an error for the ambiguity but instead automatically pick `Bedroom::repaint` and show a warning.

I do agree that seems to be manageable and I don't think it makes method resolution _cognitively_ more complex, because any such situations will result in a nice clear warning explaining the situation. It might make the _code_ significantly more complex - I'll look into it.

Reply:
OK, a concrete proposal here.

Method resolution comes up with a list of applicable `candidates`. These are currently determined by (among other things) following the chain of `Deref`s (this RFC will change this slightly to follow a chain of `Receiver`s). The number of such "autoderef" hops is recorded.

We then determine whether there's a single correct candidate, or if we should show an error.

Currently, there are two circumstances already when [the code](https://github.com/rust-lang/rust/blob/e7b22851a2c97980fd4e8235f3255af4fe0a12d5/compiler/rustc_hir_typeck/src/method/probe.rs#L1308) will favor one candidate over another:

* If some of the candidates are unstable, it will favor the others
* If there are several `impl`s of the same trait (this should not be user-visible)

If after this analysis there are _still_ several traits, it emits the errors [shown here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=218cc52ca6d8c0b584c3fa06f77b234f#:~:text=Permalink%20to%20the%20playground).

The proposal is to add a third case where it favors one of the candidates over another, if there are multiple. Specifically:

Examine the autoderef hop count on each candidate.
If one candidate has a higher autoderef hop count than all the others, pick that one (and show a warning).
Otherwise, proceed onwards to show an error.

The intention is to pick the candidate on the "innermost" type in any smart pointer, disregarding methods on the surrounding smart pointer types (but showing a warning about the problem).

Reply:
I had a go at implementing this in [this commit](https://github.com/rust-lang/rust/commit/74834988f38382ad94337b5583cc249fe522cfb7) (albeit without the warning yet). I'm pleased to say it doesn't seem to break any of the existing tests, so hopefully there aren't too many gotchas with this plan. Let me know y'alls thoughts and if nobody objects I will update the RFC to say this (and possibly broach the subject of supporting `NotNull` and `Weak` per [this discussion](https://github.com/rust-lang/rfcs/pull/3519#issuecomment-1824448078)).

Reply:
> The intention is to pick the candidate on the "innermost" type in any smart pointer,

Isn't this a breaking change because it's the opposite of the order that's already implemented and [specified in the reference](https://doc.rust-lang.org/reference/expressions/method-call-expr.html), and occasionally relied on for the “autoref/autoderef specialization” technique?

Reply:
I'm honestly not sure, but I think it's OK, because the words in the reference which we're chaniging are:

> If this results in multiple possible candidates, then it is an error

Reply:
We mustn't change what type would be picked by Deref indeed. My guess is that the rule we want would apply only in case of an ambiguity that involves a Receiver type, and would favor the Receiver method. This shouldn't be a breaking change if there isn't a Receiver involved, or if the Receiver has no inherent methods (like Box or Rc). I'd need to write it down clearly to be fully convinced that works though

Reply:
I had an example the other day: say `x: &Arc<Box<T>>` and I call `x.foo()`. Here are the methods that could be called, in my proposed order of priority ("most specific first"):
1. `foo(&Arc<Box<Self>>)` on `T`
1. `foo(&Arc<Self>)` on `Box<T>`
1. `foo(&Box<Self>)` on `T`
1. `foo(&self)` on `Arc<T>`
1. `foo(&self)` on `Box<T>`
1. `foo(&self)` on `T`
1. any trait methods

If we want "adding inherent methods to a Receiver type is not a breaking change", we need 4, 5 and 6 to be the lowest priority before trait methods. Question is: does this order break anything compared to what's allowed on stable today?

The current implementation of `arbitrary_self_types` sometimes errors on some of these combinations. The interesting cases for us is when 3 and 5 are available, or 1 and 4. This currently errors as ambiguous; we would instead pick 3 and 1 respectively.

Reply:
Ok so I did a thorough experiment (try out my madness [here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f310d10f58f72b703b22df7c9ce685ba)). It appears the algorithm picks 4 over 3 in my list above, which could appear to break my proposal. The rest of the order is ok, so today's method resolution algorithm is compatible with this ordering:

a. `foo(&Arc<Box<Self>>)` on `T`
b. `foo(&Arc<Self>)` on `Box<T>`
c. `foo(&self)` on `Arc<T>`
d. `foo(&Box<Self>)` on `T`
e. `foo(&self)` on `Box<T>`
f. `foo(&self)` on `T`

However! All is still fine. Indeed, if `Arc` was not `Deref`, we would only consider the following cases:

a. `foo(&Arc<Box<Self>>)` on `T`
b. `foo(&Arc<Self>)` on `Box<T>`
c. `foo(&self)` on `Arc<T>`

Here adding an inherent method to `Arc` would not be a breaking change. The reason adding an inherent method to `Arc` is breaking is because `Arc: Deref`, and we already know that adding methods to `Deref` types is a breaking change.

In short: if we take the current implementation of `arbitrary_self_types` and simply accept more cases that are currently errors, we can get the order above and the property that "adding inherent methods to a Receiver type is not a breaking change".

Moreover! Given that the only stable `Receiver` types don't have inherent methods (I think, right?), I believe we can even change this order a bit without breakage. It will only break users of the unstable `arbitrary_self_types` feature in that one corner case. So we can pick my initial "most specific first" order too if we prefer.

I think that proves that we're good? Did I miss anything?

Reply:
I grepped through std: the stable `Receiver`s are exactly `Box`, `Rc`, `Arc`, `Pin`, `&T`, and `&mut T`. None of these have inherent methods, except `Pin`, which has e.g. [`Pin::as_ref()`](https://doc.rust-lang.org/nightly/std/pin/struct.Pin.html#method.as_ref). So we can't change the order. But we can still do the first thing

Reply:
> @Manishearth hello, I think you were especially keen that arbitrary self types continues to support raw pointers. Do you have any views?

Missed this. Yeah, I think it's important for the ergonomics of unsafe code. Wrapper types are an okay workaround, as are free functions.

Raw pointers are basically the native rust version of "I want to be able to hold a `CppPointer<T>` without breaking it", so it would be ideal for 


I think `#[fundamental]`'s behavior may be useful for dealing with raw pointer receivers: I don't think there's anything new here that doesn't apply to `&T` as well. I think it's fine for Rust to need new editions for new methods on `*const T`; and there are some ways of making that work even without editions.